functional_programming.ipynb

```diff
@@ -0,0 +1,2835 @@
+    "```\n",
+    "def fun(a: int) -> str:\n",
+    "    return f\"{a}\"\n",
+    "```\n",
+    "\n",
+    "means that `fun` takes integer values and returns strings. Python *does not enforce* this constraint. There are tools however, which can be used to check the conistency of types in a python program."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "source": [
+    "For example, instead of doing:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "#This function greets a person by name and age\n",
+    "def greet(name: str, age: int):\n",
+    "    print(f\"Hello {name}, you are {age} years old\")\n",
+    "    \n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Hello Simone, you are 34 years old\n"
+     ]
+    }
+   ],
+   "source": [
+    "greet(\"Simone\", 34)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "source": [
+    "Because python lacks static type checking, we can do absurd things:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [
+    {
+     "ename": "NameError",
+     "evalue": "name 'greet' is not defined",
+     "output_type": "error",
+     "traceback": [
+      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
+      "Cell \u001b[0;32mIn[6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mgreet\u001b[49m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSimone\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mage\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
+      "\u001b[0;31mNameError\u001b[0m: name 'greet' is not defined"
+     ]
+    }
+   ],
+   "source": [
+    "greet(\"Simone\", \"age\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
+    "We can do better:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "\n",
+    "#This defines a simple class with some helper methods and validations\n",
+    "#in this way, we are sure that when we pass a person to the `greet_better` function, this\n",
+    "#person will have all needed attributes\n",
+    "class Person:\n",
```

Comment:
Maybe say that we will be discussing classes a bit more in detail on another day

Reply:
I would even remove the class completely and add this to the function from above

```python
if isinstance(age, int):
    ...
else:
    ...

```

Reply:
I agree,  but the points was explaining the use of types. In other functional languages (e.g rust) you would do:
```rust
struct Person{
age: i64
name: String
}
fn greet(p: Person) -> String {
 println!("Hello: {}, you are {} old", p.name, p.age);
}
```
And then you will be sure you cannot call the function with any other type. In python, you can use a class although is not really ideal.

Reply:
I added a comment in 2cb5fade8f65af31551fc7c63d9cd053bf90c97a